crutest/src/main.rs

```diff
@@ -3120,6 +3185,119 @@ async fn rand_read_write_workload(
+        cfg.blocks_per_io,
+        if cfg.blocks_per_io > 1 { "s" } else { "" },
+    );
+    print_region_description(ri, cfg.encrypted);
+    println!("----------------------------------------------");
+
+    let stop = Arc::new(AtomicBool::new(false));
+    let byte_count = Arc::new(AtomicUsize::new(0));
+
+    let block_size = ri.block_size as usize;
+    let total_blocks = ri.total_blocks;
+    let mut workers = vec![];
+    for _ in 0..cfg.io_depth {
+        let stop = stop.clone();
+        let guest = guest.clone();
+        let byte_count = byte_count.clone();
+        let handle = tokio::spawn(async move {
+            let mut buf = BytesMut::new();
+            buf.resize(cfg.blocks_per_io * block_size, 0u8);
+            let mut rng = rand_chacha::ChaCha8Rng::from_entropy();
+            rng.fill_bytes(&mut buf);
+            while !stop.load(Ordering::Acquire) {
+                let offset =
+                    rng.gen_range(0..=total_blocks - cfg.blocks_per_io);
+                guest
+                    .write(BlockIndex(offset as u64), buf.clone())
+                    .await
+                    .unwrap();
+                byte_count.fetch_add(buf.len(), Ordering::Relaxed);
+            }
+            Ok::<(), CrucibleError>(())
+        });
+        workers.push(handle);
+    }
+
+    let start = std::time::Instant::now();
+    let sleep_time = Duration::from_secs(cfg.time_secs);
+    {
+        let guest = guest.clone();
+        let stop = stop.clone();
+        tokio::spawn(async move {
+            while !stop.load(Ordering::Relaxed) {
+                let Ok(wq) = guest.query_work_queue().await else {
+                    break;
+                };
+                if sleep_time > start.elapsed() {
+                    let dt = sleep_time - start.elapsed();
+                    print!(
+                        "\r{} jobs in queue, {:.1?} secs remaining      ",
+                        wq.ds_count + wq.up_count,
+                        dt,
+                    );
+                    std::io::stdout().lock().flush().unwrap();
+                }
+                tokio::time::sleep(Duration::from_secs(1)).await;
+            }
+        });
+    }
+
+    // Sleep to fill up the buffers
+    tokio::time::sleep(Duration::from_secs(cfg.time_secs)).await;
+
+    // Create a read action
+    let read = {
+        let g = guest.clone();
+        tokio::spawn(async move {
+            let mut buffer = Buffer::new(1, block_size);
+            let start = std::time::Instant::now();
+            g.read(BlockIndex(0), &mut buffer).await.unwrap();
+            start.elapsed()
+        })
+    };
+    stop.store(true, Ordering::Release);
+
+    // Join all workers
+    for h in workers {
+        h.await??;
+    }
+
+    println!("\nawaiting final read...");
+    let read_time = read.await?;
+    println!("read took {read_time:?}");
+
+    guest.deactivate().await?;
```

Comment:
I would not deactivate here, as there is stuff that the outer loop might do after
the match on `opt.workload`.  It's likely to not be meaningful for this test, but in
general none of the workloads deactivate when they are done and this allows us
to call them from the crutest CLI.

Sometimes I'll also poke at the still running crutest with DTrace or the control 
interface to look at left over job queues or stuff like that.

Reply:
I added this deactivation because otherwise – if we call `crutest -q` to exit after the test is done – we have to go through reconciliation when `crutest` is next called.

There are two fixes that come to mind:
- `main` could deactivate the guest if `-q` is provided
- We could call `guest.flush(None).await?` here, instead of `guest.deactivate()`

I'm inclined towards the latter, because it's a smaller change.

Reply:
Yeah, adding a flush and then waiting for it to finish should avoid having to reconcile
on the next startup.

There is also a `--stable` option to crutest you can use instead of `-q` that will
wait for all jobs to finish on all downstairs before it exits.  Sometimes I do want to
"leave things dirty" and `-q` will do that for me.

But, this is your test, so you can add a flush on the end if that is what you want :)